### ISSUE TITLE
Remove .Length from MemoryManager<T>

### ISSUE THREAD

**Issue by benaadams** [dotnet/corefx#28751]
It adds no utility; and only gives scope to implementors to get `Memory` and `Length` properties out of sync.

Suggested change:
```diff
public readonly partial struct Memory<T>
{
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal Memory(MemoryManager<T> manager, int length)
+    {
+        Debug.Assert(manager != null);
+        
+        if (length < 0)
+            ThrowHelper.ThrowArgumentOutOfRangeException();
+
+        _object = manager;
+        _index = (1 << 31); // Mark as MemoryManager type
+        // Before using _index, check if _index < 0, then 'and' it with RemoveFlagsBitMask
+        _length = length;
+    }

-    public Memory(MemoryManager<T> manager, int start, int length)
-    {
-        // ...
-    }
}

public abstract partial class MemoryManager<T>
{
-    public virtual int Length => GetSpan().Length;
-    public virtual Memory<T> Memory => new Memory<T>(this, 0, Length);
+    public virtual Memory<T> Memory => CreateMemory(this, GetSpan().Length);

+    protected Memory<T> CreateMemory(int length) => new Memory<T>(this, length);
+    // Convenience api for non-zero offsets
+    protected Memory<T> CreateMemory(int start, int length) => CreateMemory(length).Slice(start);
}
```

If an implementor wants to go faster than `GetSpan().Length` rather than having a choice on overriding either `Length` or `Memory` or both; it is now clear they only need to override `Memory` as there is no alternate path.

A common usage is checking `Memory.Length` (as `.Length` isn't exposed up stream via `IMemoryOwner`) so if overriding `.Length` were the path chosen; then that's an interface call to `Memory`, which then makes a virtual call to `.Length`(then the Memory .ctor makes a further 2 virtual calls to `.Length` O_o) so its a bad path anyway.

Overriding `Memory` with this change on the other hand is would be a single interface call; then you're done, so its the better path.

e.g. the second should be chosen over the first below for performance in its common usage pattern
```diff
public class OwnedArray<T> : MemoryManager<T>
{
    T[] _array;
    int _referenceCount;
    
-   public override int Length => _array.Length;
+   public override Memory<T> Memory => CreateMemory(_array.Length)
```

Additionally currently there is the confusion of which choice to make of the two properties to override. The `Memory` override also provide greater utility as the `offset` can be changed.

cc @GrabYourPitchforks, @davidfowl, @pakrym, @davidfowl @KrzysztofCwalina

**Comment by benaadams** [comment]
cc @GrabYourPitchforks, @davidfowl, @pakrym, @davidfowl @KrzysztofCwalina

**Comment by benaadams** [comment]
**tl;dr** `.Length` is already a vestigial api; and since it hasn't shipped yet, its not too late to remove it rather than carry it forward forever.

**Comment by GrabYourPitchforks** [comment]
The `Memory<T>` ctor would query `GetSpan()` when performing bounds check validation, then? Trying to think through the whole flow graph here.

**Comment by benaadams** [comment]
> The `Memory<T>` ctor would query `GetSpan()` when performing bounds check validation, then?

Is literally no behaviour change to what its doing currently:

https://github.com/dotnet/corefx/blob/f759243d724f462da0bcef54e86588f8a55352c6/src/Common/src/CoreLib/System/Buffers/MemoryManager.cs#L13-L23

i.e.
```csharp
public abstract class MemoryManager<T> : IMemoryOwner<T>, IPinnable
{
    public virtual int Length => GetSpan().Length;
    public virtual Memory<T> Memory => new Memory<T>(this, 0, Length);
```
to
```csharp
public abstract class MemoryManager<T> : IMemoryOwner<T>, IPinnable
{
    public virtual Memory<T> Memory => new Memory<T>(this, 0, GetSpan().Length);
```

**Comment by GrabYourPitchforks** [comment]
No, I meant take a look at the code of the Memory(T) ctor that accepts a MemoryManager(T) as an input parameter. Would need to figure out how that'd change, but I don't think it'd be too bad.

**Comment by benaadams** [comment]
Missed those 2 extra virtual calls to `.Length` 😄 

Assume you trust derived `MemoryManager<T>` (e.g. for `.Length`) however you don't trust a user who manually creates a `Memory` from a `MemoryManager<T>`?

Could go for something like:
```csharp
public readonly partial struct Memory<T>
{
    internal Memory(MemoryManager<T> manager, int length)
    {
        Debug.Assert(manager != null);
        
        if (length < 0)
            ThrowHelper.ThrowArgumentOutOfRangeException();

        _object = manager;
        _index = start | (1 << 31); // Before using _index, check if _index < 0, then 'and' it with RemoveFlagsBitMask
        _length = length;
    }
    
    public Memory(MemoryManager<T> manager, int start, int length)
    {
        if (manager == null)
        {
            if (start != 0 || length != 0)
                ThrowHelper.ThrowArgumentOutOfRangeException();
            this = default;
            return; // returns default
        }

        return manager.Memory.Slice(start, length);
    }
}

public abstract partial class MemoryManager<T> : IMemoryOwner<T>, IPinnable
{
    protected Memory<T> CreateMemory(int length) => new Memory<T>(this, length);
    protected Memory<T> CreateMemory(int start, int length) => new Memory<T>(this, start, length);
    
    public virtual Memory<T> Memory => CreateMemory(this, GetSpan().Length);   
}
```
Risks a StackOverflowException if the derived `MemoryManager<T>` uses the `Memory<T>` .ctor rather than the protected `CreateMemory`; however it will always stack overflow if they do that; so easy to detect?

**Comment by benaadams** [comment]
Updated

**Comment by benaadams** [comment]
Could drop the `public Memory(MemoryManager<T> manager, int start, int length)` .ctor also; as probably should not be used? (Unless it was in `MemoryMarshall`)

```csharp
// Some aggressive inlines somewhere

public readonly partial struct Memory<T>
{
    internal Memory(MemoryManager<T> manager, int length)
    {
        Debug.Assert(manager != null);
        
        if (length < 0)
            ThrowHelper.ThrowArgumentOutOfRangeException();

        _object = manager;
        _index = start | (1 << 31); // Before using _index, check if _index < 0, then 'and' it with RemoveFlagsBitMask
        _length = length;
    }

    // Delete
    // public Memory(MemoryManager<T> manager, int start, int length)
}

public abstract partial class MemoryManager<T> : IMemoryOwner<T>, IPinnable
{
    protected Memory<T> CreateMemory(int length) => new Memory<T>(this, length);
    protected Memory<T> CreateMemory(int start, int length) => CreateMemory(length).Slice(start);
    
    public virtual Memory<T> Memory => CreateMemory(this, GetSpan().Length);   
}
```

**Comment by joshfree** [comment]
@davidfowl @pakrym what impact will this change have on asp.net repos?

**Comment by pakrym** [comment]
@joshfree Small, as far as I know, it's one file in a single repo.

**Comment by ahsonkhan** [comment]
> Small, as far as I know, it's one file in a single repo.

This should be it (couldn't find any other reference to MemoryManager in ASP.NET)
https://github.com/aspnet/Common/pull/336

**Comment by ahsonkhan** [comment]
Fixed in https://github.com/dotnet/corefx/pull/28981 / https://github.com/dotnet/corefx/pull/29041
